hack/generator/pkg/astmodel/type_merger.go

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+// TypeMerger is like a visitor for 2 types.
+//
+// Conceptually it takes (via Add) a list of functions of the form:
+//
+//     func ([ctx interface{},] a <: Type, b <: Type) (Type, error)
+//
+// where `a` and `b` can be concrete types that implement the `Type` interface.
+//
+// When `TypeMerger.Merge(Type, Type)` is invoked, it will iterate through the
+// provided functions and invoke the first one that matches the concrete types
+// passed. If none match then the fallback provided to `NewTypeMerger` will be invoked.
+//
+// The `ctx` argument can optionally be used to “smuggle” additional data down the call-chain.
+type TypeMerger struct {
+	mergers  []mergerRegistration
+	fallback MergerFunc
+}
+
+type mergerRegistration struct {
+	left, right reflect.Type
+	merge       MergerFunc
+}
+
+type MergerFunc func(ctx interface{}, left, right Type) (Type, error)
+
+func NewTypeMerger(fallback MergerFunc) TypeMerger {
+	return TypeMerger{fallback: fallback}
+}
+
+var typeInterface reflect.Type = reflect.TypeOf((*Type)(nil)).Elem() // yuck
+var errorInterface reflect.Type = reflect.TypeOf((*error)(nil)).Elem()
+var mergerFuncType reflect.Type = reflect.TypeOf((*MergerFunc)(nil)).Elem()
+
+func validateMerger(merger interface{}) (reflect.Value, reflect.Type, reflect.Type, bool) {
```

Comment:
Would a little struct rather than a 4-tuple return be clearer?

Reply:
If not, _might_ be worth defining `needsCtx` as `needsCtx := inOffset != 0` so it's at least more clear from the direct return results what that field is?